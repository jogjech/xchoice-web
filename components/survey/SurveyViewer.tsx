import React, { FunctionComponent, useState, useEffect } from "react";
import { Spin, Button, Result, Alert } from "antd";
import {
  getSurvey,
  postSurveyResponse,
  getSurveyResponse,
} from "../../apis/survey";
import { Survey } from "../../models/survey";
import QuestionViewer from "./question/QuestionViewer";
import { useRouter } from "next/router";
import styles from "./SurveyViewer.module.css";

interface Props {
  surveyId?: string;
  slug?: string;
}

const SurveyViewer: FunctionComponent<Props> = ({
  surveyId: inputSurveyId,
  slug: inputSlug,
}) => {
  // TODO: missing the handle of getSurvey and getSurveyResponse errors
  const [surveyId, setSurveyId] = useState<string>(undefined);
  const [survey, setSurvey] = useState<Survey>(undefined);
  const [loading, setLoading] = useState(true);
  const [validated, setValidated] = useState(false);
  const [selections, setSelections] = useState<number[]>([]);
  const [responseSlug, setResponseSlug] = useState<string>(undefined);
  const [posting, setPosting] = useState(false);
  const [postError, setPostError] = useState(undefined);
  const [posted, setPosted] = useState(false);
  const router = useRouter();

  useEffect(() => {
    const fetchSurvey = async () => {
      let tempSurveyId: string, tempSelections: number[];

      if (!!inputSurveyId) {
        tempSurveyId = inputSurveyId;
      } else {
        const result = await getSurveyResponse(inputSlug);
        tempSurveyId = result.surveyResponse.surveyId;
        tempSelections = result.surveyResponse.selections;
      }

      setSurveyId(tempSurveyId);

      const result = await getSurvey(tempSurveyId as string);

      setSelections(
        !!tempSelections
          ? tempSelections
          : Array.apply(null, Array(result.survey.questions.length))
      );

      setSurvey(result.survey);
      setLoading(false);
    };
    fetchSurvey();
  }, []);

  const handleSelectionChange = (questionIndex: number) => (
    newValue: number
  ) => {
    console.log(`Question ${questionIndex} received new value ${newValue}`);
    setValidated(false);
    setSelections((prevSelections) => {
      const newSelections = [...prevSelections];
      newSelections[questionIndex] = newValue;
      return newSelections;
    });
  };

  const handleSubmit = async () => {
    console.log("Submitting", selections);
    setPosting(true);
    setPostError(undefined);

    const validationPassed = validateSelections();
    if (validationPassed) {
      console.log("Validation passed");
      const result = await postSurveyResponse({
        surveyId: surveyId,
        selections: selections,
      });
      if (result.isError) {
        setPostError(result.error);
      } else {
        setResponseSlug(result.slug);
      }
      setPosted(!result.isError);
    } else {
      console.log("Validation failed");
    }
    setPosting(false);
  };

  const validateSelections: () => boolean = () => {
    setValidated(true);
    return selections.every((i) => i != null);
  };

  const handleViewResponseButtonClick = () => {
    const toPath = `/survey/response/${responseSlug}`;
    console.log(router);
    if (toPath === router.asPath) {
      router.reload();
    } else {
      router.push(`/survey/response/${responseSlug}`);
    }
  };

  return (
    <>
      {posted ? (
        <Result
          status="success"
          title={`Your response is recorded! Response id: ${responseSlug}`}
          subTitle={`Please use the response id or click the button below to view / edit your response.`}
          extra={[
            <Button
              type="primary"
              key="console"
              onClick={handleViewResponseButtonClick}
            >
              View my response
            </Button>,
          ]}
        />
      ) : loading ? (
        <div className={styles.spinContainer}>
          <Spin tip="Loading survey..."></Spin>
        </div>
      ) : (
        <div>
          <p className={styles.surveyTitle}>{survey.surveyTitle}</p>
          {survey.questions.map((question, index) => (
            <QuestionViewer
              question={question}
              key={index}
              selection={selections[index]}
              handleSelectionChange={handleSelectionChange(index)}
              submitWithNoSelection={validated && selections[index] == null}
            ></QuestionViewer>
          ))}
          <Button
            className={styles.submitButton}
            type="primary"
            onClick={handleSubmit}
            loading={posting}
          >
            Submit
          </Button>
          {postError && (
            <Alert
              message="Failed to post response. Please try again."
              type="error"
            ></Alert>
          )}
        </div>
      )}
    </>
  );
};

export default SurveyViewer;
