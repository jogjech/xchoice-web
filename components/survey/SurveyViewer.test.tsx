import React from "react";
import { shallow, ShallowWrapper } from "enzyme";
import SurveyViewer from "./SurveyViewer";

describe("Survey viewer", () => {
  let wrapperWithSurveyId: ShallowWrapper;
  let wrapperWithSlug: ShallowWrapper;
  const surveyId = "1";
  const slug = "fakljjfqklj";

  beforeEach(() => {
    wrapperWithSurveyId = shallow(
      <SurveyViewer surveyId={surveyId}></SurveyViewer>
    );
    wrapperWithSlug = shallow(<SurveyViewer slug={slug}></SurveyViewer>);
  });

  it("with survey id renders correctly", () => {
    expect(wrapperWithSurveyId).toMatchSnapshot();
  });

  it("with survey id renders spinner when loading data", () => {
    const spinner = wrapperWithSurveyId.find("Spin");
    expect(spinner.exists()).toBeTruthy();
    expect(spinner.prop("tip")).toBe("Loading survey...");
  });

  it("with slug renders correctly", () => {
    expect(wrapperWithSlug).toMatchSnapshot();
  });

  it("with slug renders spinner when loading data", () => {
    const spinner = wrapperWithSlug.find("Spin");
    expect(spinner.exists()).toBeTruthy();
    expect(spinner.prop("tip")).toBe("Loading survey...");
  });
});
