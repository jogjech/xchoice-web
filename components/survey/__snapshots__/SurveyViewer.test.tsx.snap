// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Survey viewer with slug renders correctly 1`] = `
<Fragment>
  <div
    className="spinContainer"
  >
    <Spin
      size="default"
      spinning={true}
      tip="Loading survey..."
      wrapperClassName=""
    />
  </div>
</Fragment>
`;

exports[`Survey viewer with survey id renders correctly 1`] = `
<Fragment>
  <div
    className="spinContainer"
  >
    <Spin
      size="default"
      spinning={true}
      tip="Loading survey..."
      wrapperClassName=""
    />
  </div>
</Fragment>
`;
